### ISSUE TITLE
sendTransaction based on sendToken

### ISSUE THREAD

**Issue by fadeev** [CosmWasm/cosmjs#300]
There could be a more generalized version of sendToken that sends a custom transaction. So, it would get account number, sequence and chain id, use an existing client, post request (base_req and body) to an API to get a formed tx, sign it locally and broadcast it to the chain.

Something like this:

```js
async sendTransaction({ state }, { type, body }) { // don't mind the state, it's a Vuex thing
  const { accountNumber, sequence } = await state.client.getNonce();
  const address = state.client.senderAddress;
  const chain_id = await state.client.getChainId();
  const req = {
    base_req: { chain_id, from: address },
    creator: address,
    ...body,
  };
  const { data } = await axios.post(`${API}/${chain_id}/${type}`, req);
  const { msg, fee, memo } = data.value;
  const signBytes = makeSignBytes(
    msg,
    fee,
    chain_id,
    memo,
    accountNumber,
    sequence
  );
  const signatures = [await state.wallet.sign(signBytes)];
  const signedTx = { msg: msg, fee, memo, signatures };
  return await state.client.postTx(signedTx);
}
```

Pretty useful method to have.

**Comment by webmaster128** [comment]
Very good idea.

Looking at `packages/cosmwasm/src/signingcosmwasmclient.ts` you also see a lot of repetition of this code:

```ts
    const fee = this.fees.exec;
    const { accountNumber, sequence } = await this.getNonce();
    const chainId = await this.getChainId();
    const signBytes = makeSignBytes([executeMsg], fee, chainId, memo, accountNumber, sequence);
    const signature = await this.signer.sign(this.senderAddress, signBytes);
    const signedTx: StdTx = {
      msg: [executeMsg],
      fee: fee,
      memo: memo,
      signatures: [signature],
    };

    const result = await this.postTx(signedTx);
    if (isPostTxFailure(result)) {
      throw new Error(createPostTxErrorMessage(result));
    }
```

So the same `sendTransaction` should be created in `SigningCosmWasmClient` and `SigningCosmosClient`

**Comment by ethanfrey** [comment]
I like this idea a lot, but rather than exposing this a function, I think it could be included in the OnlineSigner interface. We can provide a simple wrapper like LocalOnlineSigner that takes an LCD connection and a OfflineSigner and exposes this as a higher level method (basically what @fadeev wrote above). Or we could link to eg a chrome extension that takes the message and signs and submits it for us.

I can explain more later and continue the discussion on the OnlineSigner interface, but wanted to get feedback on this level of abstraction

**Comment by ethanfrey** [comment]
The discussion of `OfflineSigner` vs `OnlineSigner` started here: https://github.com/CosmWasm/cosmjs/issues/264 and was also inspired by some of chainapsis work on cosmosjs: https://github.com/chainapsis/cosmosjs/blob/master/src/core/walletProvider.ts#L15

The last proposal was this: https://github.com/CosmWasm/cosmjs/issues/264#issuecomment-656116845 which could have an implementation very similar to what was described above (but didn't include the post):

```ts
interface SignDoc {
  readonly account_number: string;
  readonly chain_id: string;
  readonly fee: StdFee;
  readonly memo: string;
  readonly msgs: readonly Msg[];
  readonly sequence: string;
}

interface SignResponse {
  /** The SignDoc that was signed by the user. This allows users to adjust parameters such as gas price or the memo in the signing process. Technically it allows changing everything, including the content of the messages. */
  readonly signedSignDoc: SignDoc;
  readonly signature: StdStignature;
}

export interface WalletProvider {
  // ...

  /**
   * Request signature from whichever key corresponds to provided bech32-encoded address. Rejects if not enabled.
   */
  readonly sign: (address: string, signDoc: SignDoc) => Promise<SignResponse>;
}
```

**Comment by ethanfrey** [comment]
I do like including the post, but also like the `SignDoc` argument, where most fields can be left empty and filled by the Signer:

```ts
interface SignRequest {
  // required fields
  readonly msgs: readonly Msg[];
  readonly chain_id: string;
  // optionally set by client, if unset auto-generated by the OnlineSigner before signing
  readonly account_number?: string;
  readonly fee?: StdFee;
  readonly sequence?: string;
  // fully optional (never set by signer)
  readonly memo?: string;
}

export interface OnlineSigner {
  // ...

  /**
   * Signs with whichever key corresponds to provided bech32-encoded address. Rejects if not enabled.
   * Will query chain for account_number and sequence number if not set by caller.
   * Will auto-set fee if not set by caller  
   * Will submit to the blockchain and return the TxHash if it passes `CheckTx` (else error)
   */
  readonly signAndSubmit: (address: string, request: SignRequest) => Promise<TxHash> {
    // sample implementation (from above)
      const fee = this.fees.exec;
      if request.chainId !== (await this.getChainId()) {
         throw new Error("Refuse to submit on a different chain");
      }
      let { accountNumber, sequence } = request;
      if (accountNumber === undefined || sequence === undefined) {
          { accountNumber, sequence } = await this.getNonce();
      }
      const signBytes = makeSignBytes(request.msgs, fee, request.chainId, request.memo, accountNumber, sequence);
      const signature = await this.signer.sign(address, signBytes);
      const signedTx: StdTx = {
        msg: request.msgs,
        fee: fee,
        memo: request.memo,
        signatures: [signature],
    };

    // use mode = SYNC to only wait for CheckTX not block inclusion
    const result = await this.postTx(signedTx, SYNC);
    if (isPostTxFailure(result)) {
      throw new Error(createPostTxErrorMessage(result));
    }
    return result.txHash;
  }
}

```

**Comment by webmaster128** [comment]
Thank you Ethan! I agree this is a more general solution to the problem. However, I don't think we will get it sufficiently refned and implemented this week and in 0.22.0. I created a new ticket here: https://github.com/CosmWasm/cosmjs/issues/316. I'll close this ticket with a smaller change of the current code.
